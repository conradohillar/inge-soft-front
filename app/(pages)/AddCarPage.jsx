import { Text } from 'react-native'
import React, { useState } from 'react'
import { SafeAreaView } from 'react-native-safe-area-context'
import { YStack, XStack, ScrollView } from 'tamagui'
import CustomInput from '../../components/CustomInput'
import ButtonNext from '../../components/ButtonNext'
import { Link, useRouter } from 'expo-router'
import LoadingPage from '../(pages)/LoadingPage'
import { useMutation, useQueryClient } from '@tanstack/react-query'
import ErrorPage from '../(pages)/ErrorPage'
import { icons } from '../../constants'
import { TouchableWithoutFeedback, Keyboard, KeyboardAvoidingView, Platform } from 'react-native'
import { newCar } from '../../services/users'




export default function AddCarPage() {
  const queryClient = useQueryClient();
  const [model, setModel] = useState('')
  const [plate, setPlate] = useState('')
  const [color, setColor] = useState('')

  const mutation = useMutation({
    mutationFn: (carData) => newCar(carData),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['getCars'] })
      queryClient.invalidateQueries({ queryKey: ['getRideData'] })
      router.push({
        pathname: "/(pages)/PostSuccessful",
        params: {
          title: "Agregaste tu auto!",
          section: "Mis autos",
          sectionSource: icons.car,
          returnTo: "Volver a Mis autos",
          returnToSource: icons.car,
          returnToRef: "/(pages)/MyCarsPage"
        }
      });
    }
  })


  const router = useRouter();

  const handleContinue = () => {
    const obj = {
      "model": model,
      "plate": plate,
      "color": color
    }

    mutation.mutate(obj);
  };


  if (mutation.isPending) {
    return <LoadingPage />;
  }

  if (mutation.isError) {
    return <ErrorPage />;
  }


  return (
    <KeyboardAvoidingView
      behavior={Platform.OS === "ios" ? "padding" : "height"} // "padding" para iOS, "height" para Android
      style={{ flex: 1 }}
    >
      <TouchableWithoutFeedback onPress={Keyboard.dismiss} className={"h-full"}>
        <SafeAreaView className="bg-background h-full w-full">
          <ScrollView className="h-full">
          <YStack className="h-full justify-evenly">
            <YStack className="items-center justify-center">
              <Text className="text-black text-4xl font-qbold">Carg√° los datos</Text>
              <Text className="text-black text-4xl font-qbold">de
                <Text className="text-primary text-4xl font-qbold"> tu auto</Text>
              </Text>
            </YStack>
            <YStack className="items-center justify-center">
              <CustomInput title="Modelo" value={model} handleChangeText={setModel} />
              <CustomInput title="Patente" value={plate} handleChangeText={setPlate} />
              <CustomInput title="Color" value={color} handleChangeText={setColor} />
            </YStack>
            <YStack className="items-center">
              <ButtonNext onPress={handleContinue}>
                <Text className="text-white text-xl font-qsemibold">Agregar auto</Text>
              </ButtonNext>
              <Link href="/(pages)/MyCarsPage" asChild>
                <Text className="text-base text-primary font-qsemibold underline">Volver</Text>
              </Link>
            </YStack>
          </YStack>
          </ScrollView>
        </SafeAreaView>
      </TouchableWithoutFeedback>
    </KeyboardAvoidingView>
  )
}
