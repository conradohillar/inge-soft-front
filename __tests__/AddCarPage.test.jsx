import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react-native';
import AddCarPage from '../app/(pages)/AddCarPage';
import axios from 'axios';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { TamaguiProvider } from 'tamagui';
import config from '../tamagui.config'

jest.mock('axios');

const queryClient = new QueryClient();
const Provider = ({ children }) => (
    <TamaguiProvider config={config}>
        <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </TamaguiProvider>
);

describe('AddCarPage', () => {
    beforeEach(() => {
        axios.post.mockClear();
    });

    it('renders correctly', () => {
        const { getByText } = render(
            <Provider>
                <AddCarPage />
            </Provider>
        );
        expect(getByText('Cargá los datos')).toBeTruthy();
        expect(getByText('tu auto')).toBeTruthy();
    });

    it('displays error messages when form is submitted with empty fields', async () => {
        const { getByText } = render(
            <Provider>
                <AddCarPage />
            </Provider>
        );
        fireEvent.press(getByText('Agregar auto'));

        await waitFor(() => {
            expect(getByText('Model is required')).toBeTruthy();
            expect(getByText('Plate is required')).toBeTruthy();
            expect(getByText('Color is required')).toBeTruthy();
        });
    });

    it('calls mutation with correct data when form is submitted', async () => {
        axios.post.mockResolvedValueOnce({ data: {} });

        const { getByText, getByPlaceholderText } = render(
            <Provider>
                <AddCarPage />
            </Provider>
        );

        fireEvent.changeText(getByPlaceholderText('Ej: Toyota etios'), 'Toyota etios');
        fireEvent.changeText(getByPlaceholderText('Ej: abc-123'), 'abc-123');
        fireEvent.changeText(getByPlaceholderText('Ej: Rojo'), 'Red');
        fireEvent.press(getByText('Agregar auto'));

        await waitFor(() => {
            expect(axios.post).toHaveBeenCalledWith('http://192.168.100.158:8000/users/addcar', {
                model: 'Toyota etios',
                plate: 'abc-123',
                color: 'Red',
            },
                { "headers": { "Authorization": "Bearer null", "Content-Type": "application/json" } });
        });
    });
    it('updates form field values correctly when typing', async () => {
        const { getByPlaceholderText } = render(
            <Provider>
                <AddCarPage />
            </Provider>
        );


        const modelInput = getByPlaceholderText('Ej: Toyota etios')
        const plateInput = getByPlaceholderText('Ej: abc-123');
        const colorInput = getByPlaceholderText('Ej: Rojo');


        fireEvent.changeText(modelInput, 'Toyota');
        fireEvent.changeText(plateInput, 'abc-123');
        fireEvent.changeText(colorInput, 'Red');


        await waitFor(() => {
            expect(modelInput.props.value).toBe('Toyota');
            expect(plateInput.props.value).toBe('abc-123');
            expect(colorInput.props.value).toBe('Red');
        });
    });


    describe('add car page Snapshot', () => {
        it('debería coincidir con la snapshot', () => {
            const tree = render(<Provider><AddCarPage /></Provider>).toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});
