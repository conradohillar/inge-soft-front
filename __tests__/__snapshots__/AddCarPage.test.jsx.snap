// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCarPage add car page Snapshot debería coincidir con la snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RNCSafeAreaView
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
        },
        {
          "height": "100%",
        },
        {
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "height": "100%",
          "justifyContent": "space-evenly",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              {
                "fontSize": 36,
                "lineHeight": 40,
              },
              {
                "fontFamily": "Quicksand-Bold",
              },
            ]
          }
        >
          Cargá los datos
        </Text>
        <Text
          style={
            [
              {
                "color": "#000",
              },
              {
                "fontSize": 36,
                "lineHeight": 40,
              },
              {
                "fontFamily": "Quicksand-Bold",
              },
            ]
          }
        >
          de
          <Text
            style={
              [
                {
                  "color": "#59A58A",
                },
                {
                  "fontSize": 36,
                  "lineHeight": 40,
                },
                {
                  "fontFamily": "Quicksand-Bold",
                },
              ]
            }
          >
             
            tu auto
          </Text>
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "flex-start",
              "flexDirection": "column",
              "justifyContent": "center",
              "marginBottom": 8,
              "marginLeft": 8,
              "marginRight": 8,
              "marginTop": 8,
              "width": "100%",
            }
          }
        >
          <View
            style={
              [
                {
                  "width": "100%",
                },
                {
                  "alignItems": "center",
                },
              ]
            }
          >
            <View
              style={
                [
                  {
                    "width": "85%",
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Quicksand-Bold",
                    },
                    {
                      "color": "#4b5563",
                    },
                    {
                      "paddingLeft": 6,
                      "paddingRight": 6,
                    },
                    {
                      "marginBottom": 8,
                    },
                  ]
                }
              >
                Modelo
              </Text>
            </View>
          </View>
          <View
            style={
              [
                {
                  "width": "100%",
                },
                {
                  "alignItems": "center",
                },
              ]
            }
          >
            <TextInput
              backgroundColor="#EEE"
              color="black"
              editable={true}
              multiline={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              p={15}
              placeholder="Ej: Toyota etios"
              placeholderTextColor="#888"
              scrollEnabled={true}
              secureTextEntry={false}
              style={
                [
                  {
                    "width": "85%",
                  },
                  {
                    "backgroundColor": "#EEE",
                  },
                  {
                    "borderBottomLeftRadius": 8,
                    "borderBottomRightRadius": 8,
                    "borderTopLeftRadius": 8,
                    "borderTopRightRadius": 8,
                  },
                  {
                    "color": "#000",
                  },
                  {
                    "paddingBottom": 16,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "paddingTop": 16,
                  },
                  {
                    "borderBottomWidth": 1,
                    "borderLeftWidth": 1,
                    "borderRightWidth": 1,
                    "borderTopWidth": 1,
                  },
                  {
                    "borderBottomColor": "#111827",
                    "borderLeftColor": "#111827",
                    "borderRightColor": "#111827",
                    "borderTopColor": "#111827",
                  },
                ]
              }
              textAlignVertical="top"
              value=""
            />
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "flex-start",
              "flexDirection": "column",
              "justifyContent": "center",
              "marginBottom": 8,
              "marginLeft": 8,
              "marginRight": 8,
              "marginTop": 8,
              "width": "100%",
            }
          }
        >
          <View
            style={
              [
                {
                  "width": "100%",
                },
                {
                  "alignItems": "center",
                },
              ]
            }
          >
            <View
              style={
                [
                  {
                    "width": "85%",
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Quicksand-Bold",
                    },
                    {
                      "color": "#4b5563",
                    },
                    {
                      "paddingLeft": 6,
                      "paddingRight": 6,
                    },
                    {
                      "marginBottom": 8,
                    },
                  ]
                }
              >
                Patente
              </Text>
            </View>
          </View>
          <View
            style={
              [
                {
                  "width": "100%",
                },
                {
                  "alignItems": "center",
                },
              ]
            }
          >
            <TextInput
              backgroundColor="#EEE"
              color="black"
              editable={true}
              multiline={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              p={15}
              placeholder="Ej: abc-123"
              placeholderTextColor="#888"
              scrollEnabled={true}
              secureTextEntry={false}
              style={
                [
                  {
                    "width": "85%",
                  },
                  {
                    "backgroundColor": "#EEE",
                  },
                  {
                    "borderBottomLeftRadius": 8,
                    "borderBottomRightRadius": 8,
                    "borderTopLeftRadius": 8,
                    "borderTopRightRadius": 8,
                  },
                  {
                    "color": "#000",
                  },
                  {
                    "paddingBottom": 16,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "paddingTop": 16,
                  },
                  {
                    "borderBottomWidth": 1,
                    "borderLeftWidth": 1,
                    "borderRightWidth": 1,
                    "borderTopWidth": 1,
                  },
                  {
                    "borderBottomColor": "#111827",
                    "borderLeftColor": "#111827",
                    "borderRightColor": "#111827",
                    "borderTopColor": "#111827",
                  },
                ]
              }
              textAlignVertical="top"
              value=""
            />
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "flex-start",
              "flexDirection": "column",
              "justifyContent": "center",
              "marginBottom": 8,
              "marginLeft": 8,
              "marginRight": 8,
              "marginTop": 8,
              "width": "100%",
            }
          }
        >
          <View
            style={
              [
                {
                  "width": "100%",
                },
                {
                  "alignItems": "center",
                },
              ]
            }
          >
            <View
              style={
                [
                  {
                    "width": "85%",
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Quicksand-Bold",
                    },
                    {
                      "color": "#4b5563",
                    },
                    {
                      "paddingLeft": 6,
                      "paddingRight": 6,
                    },
                    {
                      "marginBottom": 8,
                    },
                  ]
                }
              >
                Color
              </Text>
            </View>
          </View>
          <View
            style={
              [
                {
                  "width": "100%",
                },
                {
                  "alignItems": "center",
                },
              ]
            }
          >
            <TextInput
              backgroundColor="#EEE"
              color="black"
              editable={true}
              multiline={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              p={15}
              placeholder="Ej: Rojo"
              placeholderTextColor="#888"
              scrollEnabled={true}
              secureTextEntry={false}
              style={
                [
                  {
                    "width": "85%",
                  },
                  {
                    "backgroundColor": "#EEE",
                  },
                  {
                    "borderBottomLeftRadius": 8,
                    "borderBottomRightRadius": 8,
                    "borderTopLeftRadius": 8,
                    "borderTopRightRadius": 8,
                  },
                  {
                    "color": "#000",
                  },
                  {
                    "paddingBottom": 16,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "paddingTop": 16,
                  },
                  {
                    "borderBottomWidth": 1,
                    "borderLeftWidth": 1,
                    "borderRightWidth": 1,
                    "borderTopWidth": 1,
                  },
                  {
                    "borderBottomColor": "#111827",
                    "borderLeftColor": "#111827",
                    "borderRightColor": "#111827",
                    "borderTopColor": "#111827",
                  },
                ]
              }
              textAlignVertical="top"
              value=""
            />
          </View>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "column",
          }
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
              },
              {
                "width": "100%",
              },
            ]
          }
        >
          <View
            focusVisibleStyle={{}}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            role="button"
            style={
              {
                "alignItems": "center",
                "backgroundColor": "black",
                "borderBottomColor": "black",
                "borderBottomLeftRadius": 9999,
                "borderBottomRightRadius": 9999,
                "borderBottomWidth": 1,
                "borderColor": "black",
                "borderLeftColor": "black",
                "borderLeftWidth": 1,
                "borderRightColor": "black",
                "borderRightWidth": 1,
                "borderStyle": "solid",
                "borderTopColor": "black",
                "borderTopLeftRadius": 9999,
                "borderTopRightRadius": 9999,
                "borderTopWidth": 1,
                "borderWidth": 1,
                "flexDirection": "row",
                "flexWrap": "nowrap",
                "height": 75,
                "justifyContent": "center",
                "marginBottom": 8,
                "marginLeft": 8,
                "marginRight": 8,
                "marginTop": 8,
                "paddingLeft": 18,
                "paddingRight": 18,
                "width": "85%",
              }
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#fff",
                  },
                  {
                    "fontSize": 20,
                    "lineHeight": 28,
                  },
                  {
                    "fontFamily": "Quicksand-SemiBold",
                  },
                ]
              }
            >
              Agregar auto
            </Text>
          </View>
        </View>
        <Text
          href="/MyCarsPage"
          onPress={[Function]}
          role="link"
          style={
            [
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
              {
                "color": "#59A58A",
              },
              {
                "fontFamily": "Quicksand-SemiBold",
              },
              {
                "textDecorationLine": "underline",
              },
            ]
          }
        >
          Volver
        </Text>
      </View>
    </View>
  </RNCSafeAreaView>
</View>
`;
